eval/src/vespa/eval/eval/tensor_plans.h

```diff
@@ -0,0 +1,77 @@
+// Copyright Verizon Media. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
+
+#pragma once
+
+#include "value_type.h"
+#include <vespa/vespalib/stllike/string.h>
+#include <vector>
+
+namespace vespalib::eval {
+
+class ValueBuilderFactory;
+
+//-----------------------------------------------------------------------------
+
+/**
+ * Plan for how to traverse two partially overlapping dense subspaces
+ * in parallel, identifying all matching cell index combinations, in
+ * the exact order the joined cells will be stored in the result. The
+ * plan can be made up-front during tensor function compilation.
+ **/
+struct DenseJoinPlan {
+    size_t lhs_size;
+    size_t rhs_size;
+    size_t out_size;
+    std::vector<size_t> loop_cnt;
+    std::vector<size_t> lhs_stride;
+    std::vector<size_t> rhs_stride;
+    DenseJoinPlan(const ValueType &lhs_type, const ValueType &rhs_type);
+    ~DenseJoinPlan();
+    template <typename F> void execute(size_t lhs, size_t rhs, F &&f) const {
```

Comment:
is it possible to move the various execute() implementations to the .cpp file? or would that lead to less inlining opportunities?

Reply:
I think we want it in a separate header, for cross-use in different plans. We do not want to call the function via pointer, so I think we should avoid having it in a cpp file.